<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Source of the Rust file `src/rna_counts.rs`."><title>rna_counts.rs - source</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-f3501f0f5ae15dfb.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="sumi" data-themes="" data-resource-suffix="" data-rustdoc-version="1.71.1 (eb26296b5 2023-08-03)" data-search-js="search-4926e5fc22a5646a.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../../static.files/source-script-11255107d0a65077.js"></script><script defer src="../../source-files.js"></script><script defer src="../../static.files/main-f0540c1d82cde29b.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc source"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"></nav><main><nav class="sub"><a class="sub-logo-container" href="../../sumi/index.html"><img class="rust-logo" src="../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="example-wrap"><div data-nosnippet><pre class="src-line-numbers"><a href="#1" id="1">1</a>
<a href="#2" id="2">2</a>
<a href="#3" id="3">3</a>
<a href="#4" id="4">4</a>
<a href="#5" id="5">5</a>
<a href="#6" id="6">6</a>
<a href="#7" id="7">7</a>
<a href="#8" id="8">8</a>
<a href="#9" id="9">9</a>
<a href="#10" id="10">10</a>
<a href="#11" id="11">11</a>
<a href="#12" id="12">12</a>
<a href="#13" id="13">13</a>
<a href="#14" id="14">14</a>
<a href="#15" id="15">15</a>
<a href="#16" id="16">16</a>
<a href="#17" id="17">17</a>
<a href="#18" id="18">18</a>
<a href="#19" id="19">19</a>
<a href="#20" id="20">20</a>
<a href="#21" id="21">21</a>
<a href="#22" id="22">22</a>
<a href="#23" id="23">23</a>
<a href="#24" id="24">24</a>
<a href="#25" id="25">25</a>
<a href="#26" id="26">26</a>
<a href="#27" id="27">27</a>
<a href="#28" id="28">28</a>
<a href="#29" id="29">29</a>
<a href="#30" id="30">30</a>
<a href="#31" id="31">31</a>
<a href="#32" id="32">32</a>
<a href="#33" id="33">33</a>
<a href="#34" id="34">34</a>
<a href="#35" id="35">35</a>
<a href="#36" id="36">36</a>
<a href="#37" id="37">37</a>
<a href="#38" id="38">38</a>
<a href="#39" id="39">39</a>
<a href="#40" id="40">40</a>
<a href="#41" id="41">41</a>
<a href="#42" id="42">42</a>
<a href="#43" id="43">43</a>
<a href="#44" id="44">44</a>
<a href="#45" id="45">45</a>
<a href="#46" id="46">46</a>
<a href="#47" id="47">47</a>
<a href="#48" id="48">48</a>
<a href="#49" id="49">49</a>
<a href="#50" id="50">50</a>
<a href="#51" id="51">51</a>
<a href="#52" id="52">52</a>
<a href="#53" id="53">53</a>
<a href="#54" id="54">54</a>
<a href="#55" id="55">55</a>
<a href="#56" id="56">56</a>
<a href="#57" id="57">57</a>
<a href="#58" id="58">58</a>
<a href="#59" id="59">59</a>
<a href="#60" id="60">60</a>
<a href="#61" id="61">61</a>
<a href="#62" id="62">62</a>
<a href="#63" id="63">63</a>
<a href="#64" id="64">64</a>
<a href="#65" id="65">65</a>
<a href="#66" id="66">66</a>
<a href="#67" id="67">67</a>
<a href="#68" id="68">68</a>
<a href="#69" id="69">69</a>
<a href="#70" id="70">70</a>
<a href="#71" id="71">71</a>
<a href="#72" id="72">72</a>
<a href="#73" id="73">73</a>
<a href="#74" id="74">74</a>
<a href="#75" id="75">75</a>
<a href="#76" id="76">76</a>
<a href="#77" id="77">77</a>
<a href="#78" id="78">78</a>
<a href="#79" id="79">79</a>
<a href="#80" id="80">80</a>
<a href="#81" id="81">81</a>
<a href="#82" id="82">82</a>
<a href="#83" id="83">83</a>
<a href="#84" id="84">84</a>
<a href="#85" id="85">85</a>
<a href="#86" id="86">86</a>
<a href="#87" id="87">87</a>
<a href="#88" id="88">88</a>
<a href="#89" id="89">89</a>
<a href="#90" id="90">90</a>
<a href="#91" id="91">91</a>
<a href="#92" id="92">92</a>
<a href="#93" id="93">93</a>
<a href="#94" id="94">94</a>
<a href="#95" id="95">95</a>
<a href="#96" id="96">96</a>
<a href="#97" id="97">97</a>
<a href="#98" id="98">98</a>
<a href="#99" id="99">99</a>
<a href="#100" id="100">100</a>
<a href="#101" id="101">101</a>
<a href="#102" id="102">102</a>
<a href="#103" id="103">103</a>
<a href="#104" id="104">104</a>
<a href="#105" id="105">105</a>
<a href="#106" id="106">106</a>
<a href="#107" id="107">107</a>
<a href="#108" id="108">108</a>
<a href="#109" id="109">109</a>
<a href="#110" id="110">110</a>
<a href="#111" id="111">111</a>
<a href="#112" id="112">112</a>
<a href="#113" id="113">113</a>
<a href="#114" id="114">114</a>
<a href="#115" id="115">115</a>
<a href="#116" id="116">116</a>
<a href="#117" id="117">117</a>
<a href="#118" id="118">118</a>
<a href="#119" id="119">119</a>
<a href="#120" id="120">120</a>
<a href="#121" id="121">121</a>
<a href="#122" id="122">122</a>
<a href="#123" id="123">123</a>
<a href="#124" id="124">124</a>
<a href="#125" id="125">125</a>
<a href="#126" id="126">126</a>
<a href="#127" id="127">127</a>
<a href="#128" id="128">128</a>
<a href="#129" id="129">129</a>
<a href="#130" id="130">130</a>
<a href="#131" id="131">131</a>
<a href="#132" id="132">132</a>
<a href="#133" id="133">133</a>
<a href="#134" id="134">134</a>
<a href="#135" id="135">135</a>
<a href="#136" id="136">136</a>
<a href="#137" id="137">137</a>
<a href="#138" id="138">138</a>
<a href="#139" id="139">139</a>
<a href="#140" id="140">140</a>
<a href="#141" id="141">141</a>
<a href="#142" id="142">142</a>
<a href="#143" id="143">143</a>
<a href="#144" id="144">144</a>
<a href="#145" id="145">145</a>
<a href="#146" id="146">146</a>
<a href="#147" id="147">147</a>
<a href="#148" id="148">148</a>
<a href="#149" id="149">149</a>
<a href="#150" id="150">150</a>
<a href="#151" id="151">151</a>
<a href="#152" id="152">152</a>
<a href="#153" id="153">153</a>
<a href="#154" id="154">154</a>
<a href="#155" id="155">155</a>
<a href="#156" id="156">156</a>
<a href="#157" id="157">157</a>
<a href="#158" id="158">158</a>
<a href="#159" id="159">159</a>
<a href="#160" id="160">160</a>
<a href="#161" id="161">161</a>
<a href="#162" id="162">162</a>
<a href="#163" id="163">163</a>
<a href="#164" id="164">164</a>
<a href="#165" id="165">165</a>
<a href="#166" id="166">166</a>
<a href="#167" id="167">167</a>
<a href="#168" id="168">168</a>
<a href="#169" id="169">169</a>
<a href="#170" id="170">170</a>
<a href="#171" id="171">171</a>
<a href="#172" id="172">172</a>
<a href="#173" id="173">173</a>
<a href="#174" id="174">174</a>
<a href="#175" id="175">175</a>
<a href="#176" id="176">176</a>
<a href="#177" id="177">177</a>
<a href="#178" id="178">178</a>
<a href="#179" id="179">179</a>
<a href="#180" id="180">180</a>
<a href="#181" id="181">181</a>
<a href="#182" id="182">182</a>
<a href="#183" id="183">183</a>
<a href="#184" id="184">184</a>
<a href="#185" id="185">185</a>
<a href="#186" id="186">186</a>
<a href="#187" id="187">187</a>
<a href="#188" id="188">188</a>
<a href="#189" id="189">189</a>
<a href="#190" id="190">190</a>
<a href="#191" id="191">191</a>
<a href="#192" id="192">192</a>
<a href="#193" id="193">193</a>
<a href="#194" id="194">194</a>
<a href="#195" id="195">195</a>
<a href="#196" id="196">196</a>
<a href="#197" id="197">197</a>
<a href="#198" id="198">198</a>
<a href="#199" id="199">199</a>
<a href="#200" id="200">200</a>
<a href="#201" id="201">201</a>
<a href="#202" id="202">202</a>
<a href="#203" id="203">203</a>
<a href="#204" id="204">204</a>
<a href="#205" id="205">205</a>
<a href="#206" id="206">206</a>
<a href="#207" id="207">207</a>
<a href="#208" id="208">208</a>
<a href="#209" id="209">209</a>
<a href="#210" id="210">210</a>
<a href="#211" id="211">211</a>
<a href="#212" id="212">212</a>
<a href="#213" id="213">213</a>
<a href="#214" id="214">214</a>
<a href="#215" id="215">215</a>
<a href="#216" id="216">216</a>
<a href="#217" id="217">217</a>
<a href="#218" id="218">218</a>
<a href="#219" id="219">219</a>
<a href="#220" id="220">220</a>
<a href="#221" id="221">221</a>
<a href="#222" id="222">222</a>
<a href="#223" id="223">223</a>
<a href="#224" id="224">224</a>
<a href="#225" id="225">225</a>
<a href="#226" id="226">226</a>
<a href="#227" id="227">227</a>
<a href="#228" id="228">228</a>
<a href="#229" id="229">229</a>
<a href="#230" id="230">230</a>
<a href="#231" id="231">231</a>
<a href="#232" id="232">232</a>
<a href="#233" id="233">233</a>
<a href="#234" id="234">234</a>
<a href="#235" id="235">235</a>
<a href="#236" id="236">236</a>
<a href="#237" id="237">237</a>
<a href="#238" id="238">238</a>
<a href="#239" id="239">239</a>
<a href="#240" id="240">240</a>
<a href="#241" id="241">241</a>
<a href="#242" id="242">242</a>
<a href="#243" id="243">243</a>
<a href="#244" id="244">244</a>
<a href="#245" id="245">245</a>
<a href="#246" id="246">246</a>
<a href="#247" id="247">247</a>
<a href="#248" id="248">248</a>
<a href="#249" id="249">249</a>
<a href="#250" id="250">250</a>
<a href="#251" id="251">251</a>
<a href="#252" id="252">252</a>
<a href="#253" id="253">253</a>
<a href="#254" id="254">254</a>
<a href="#255" id="255">255</a>
<a href="#256" id="256">256</a>
<a href="#257" id="257">257</a>
<a href="#258" id="258">258</a>
<a href="#259" id="259">259</a>
<a href="#260" id="260">260</a>
<a href="#261" id="261">261</a>
<a href="#262" id="262">262</a>
<a href="#263" id="263">263</a>
<a href="#264" id="264">264</a>
<a href="#265" id="265">265</a>
<a href="#266" id="266">266</a>
<a href="#267" id="267">267</a>
<a href="#268" id="268">268</a>
<a href="#269" id="269">269</a>
<a href="#270" id="270">270</a>
<a href="#271" id="271">271</a>
<a href="#272" id="272">272</a>
<a href="#273" id="273">273</a>
<a href="#274" id="274">274</a>
<a href="#275" id="275">275</a>
<a href="#276" id="276">276</a>
<a href="#277" id="277">277</a>
<a href="#278" id="278">278</a>
<a href="#279" id="279">279</a>
<a href="#280" id="280">280</a>
<a href="#281" id="281">281</a>
<a href="#282" id="282">282</a>
<a href="#283" id="283">283</a>
<a href="#284" id="284">284</a>
<a href="#285" id="285">285</a>
<a href="#286" id="286">286</a>
<a href="#287" id="287">287</a>
<a href="#288" id="288">288</a>
<a href="#289" id="289">289</a>
<a href="#290" id="290">290</a>
<a href="#291" id="291">291</a>
<a href="#292" id="292">292</a>
<a href="#293" id="293">293</a>
<a href="#294" id="294">294</a>
<a href="#295" id="295">295</a>
<a href="#296" id="296">296</a>
<a href="#297" id="297">297</a>
<a href="#298" id="298">298</a>
<a href="#299" id="299">299</a>
<a href="#300" id="300">300</a>
<a href="#301" id="301">301</a>
<a href="#302" id="302">302</a>
<a href="#303" id="303">303</a>
<a href="#304" id="304">304</a>
<a href="#305" id="305">305</a>
<a href="#306" id="306">306</a>
<a href="#307" id="307">307</a>
<a href="#308" id="308">308</a>
<a href="#309" id="309">309</a>
<a href="#310" id="310">310</a>
<a href="#311" id="311">311</a>
<a href="#312" id="312">312</a>
<a href="#313" id="313">313</a>
<a href="#314" id="314">314</a>
<a href="#315" id="315">315</a>
<a href="#316" id="316">316</a>
<a href="#317" id="317">317</a>
<a href="#318" id="318">318</a>
<a href="#319" id="319">319</a>
<a href="#320" id="320">320</a>
<a href="#321" id="321">321</a>
<a href="#322" id="322">322</a>
<a href="#323" id="323">323</a>
<a href="#324" id="324">324</a>
<a href="#325" id="325">325</a>
<a href="#326" id="326">326</a>
<a href="#327" id="327">327</a>
<a href="#328" id="328">328</a>
<a href="#329" id="329">329</a>
<a href="#330" id="330">330</a>
<a href="#331" id="331">331</a>
<a href="#332" id="332">332</a>
<a href="#333" id="333">333</a>
<a href="#334" id="334">334</a>
<a href="#335" id="335">335</a>
<a href="#336" id="336">336</a>
<a href="#337" id="337">337</a>
<a href="#338" id="338">338</a>
<a href="#339" id="339">339</a>
<a href="#340" id="340">340</a>
<a href="#341" id="341">341</a>
<a href="#342" id="342">342</a>
<a href="#343" id="343">343</a>
<a href="#344" id="344">344</a>
<a href="#345" id="345">345</a>
<a href="#346" id="346">346</a>
<a href="#347" id="347">347</a>
<a href="#348" id="348">348</a>
<a href="#349" id="349">349</a>
<a href="#350" id="350">350</a>
<a href="#351" id="351">351</a>
<a href="#352" id="352">352</a>
<a href="#353" id="353">353</a>
<a href="#354" id="354">354</a>
<a href="#355" id="355">355</a>
<a href="#356" id="356">356</a>
<a href="#357" id="357">357</a>
<a href="#358" id="358">358</a>
<a href="#359" id="359">359</a>
<a href="#360" id="360">360</a>
<a href="#361" id="361">361</a>
<a href="#362" id="362">362</a>
<a href="#363" id="363">363</a>
<a href="#364" id="364">364</a>
<a href="#365" id="365">365</a>
<a href="#366" id="366">366</a>
<a href="#367" id="367">367</a>
<a href="#368" id="368">368</a>
<a href="#369" id="369">369</a>
<a href="#370" id="370">370</a>
<a href="#371" id="371">371</a>
<a href="#372" id="372">372</a>
<a href="#373" id="373">373</a>
<a href="#374" id="374">374</a>
<a href="#375" id="375">375</a>
<a href="#376" id="376">376</a>
<a href="#377" id="377">377</a>
<a href="#378" id="378">378</a>
<a href="#379" id="379">379</a>
<a href="#380" id="380">380</a>
<a href="#381" id="381">381</a>
<a href="#382" id="382">382</a>
<a href="#383" id="383">383</a>
<a href="#384" id="384">384</a>
<a href="#385" id="385">385</a>
<a href="#386" id="386">386</a>
<a href="#387" id="387">387</a>
<a href="#388" id="388">388</a>
<a href="#389" id="389">389</a>
<a href="#390" id="390">390</a>
<a href="#391" id="391">391</a>
<a href="#392" id="392">392</a>
<a href="#393" id="393">393</a>
<a href="#394" id="394">394</a>
<a href="#395" id="395">395</a>
<a href="#396" id="396">396</a>
<a href="#397" id="397">397</a>
<a href="#398" id="398">398</a>
<a href="#399" id="399">399</a>
<a href="#400" id="400">400</a>
<a href="#401" id="401">401</a>
<a href="#402" id="402">402</a>
<a href="#403" id="403">403</a>
<a href="#404" id="404">404</a>
<a href="#405" id="405">405</a>
<a href="#406" id="406">406</a>
<a href="#407" id="407">407</a>
<a href="#408" id="408">408</a>
<a href="#409" id="409">409</a>
<a href="#410" id="410">410</a>
</pre></div><pre class="rust"><code><span class="kw">use crate</span>::{ProgressBar, ProgressStyle};
<span class="kw">use </span><span class="kw">crate</span>::Path;
<span class="kw">use </span><span class="kw">crate</span>::Command;
<span class="kw">use crate</span>::{HashSet, HashMap};
<span class="kw">use </span><span class="kw">crate</span>::graph::find_true_umis;
<span class="kw">use </span><span class="kw">crate</span>::graph::deduplicate_bam;
<span class="kw">use </span><span class="kw">crate</span>::capture_target_files;
<span class="kw">use </span><span class="kw">crate</span>::quality::average_read_quality;
<span class="kw">use </span><span class="kw">crate</span>::get_read_counts;
<span class="kw">use </span><span class="kw">crate</span>::Config;
<span class="kw">use </span><span class="kw">crate</span>::File;
<span class="kw">use </span><span class="kw">crate</span>::BufReader;
<span class="kw">use </span><span class="kw">crate</span>::DataFrame;
<span class="kw">use </span><span class="kw">crate</span>::Series;
<span class="kw">use </span><span class="kw">crate</span>::CsvWriter;
<span class="kw">use </span><span class="kw">crate</span>::Error;
<span class="kw">use </span><span class="kw">crate</span>::csv_writer;
<span class="kw">use </span>std::io::BufRead;
<span class="kw">use </span>polars::prelude::{NamedFrom, SerWriter};
<span class="kw">use </span>std::io;

<span class="comment">// Define a custom error type that can represent both io::Error and bio::io::fastq::Error
// Can be used to determine if an error occurred when reading the input or the fastq file
</span><span class="attr">#[derive(Error, Debug)]
</span><span class="kw">pub enum </span>MyError {
    <span class="attr">#[error(<span class="string">&quot;I/O error&quot;</span>)]
    </span>Io(<span class="attr">#[from] </span>io::Error),
    <span class="attr">#[error(<span class="string">&quot;FASTQ error&quot;</span>)]
    </span>Fastq(<span class="attr">#[from] </span>bio::io::fastq::Error),
}

<span class="doccomment">/// Checks a SAM file to determine if it contains alignment data.
///
/// The function reads through the SAM file line by line, looking for any line that does 
/// not start with &#39;@&#39;, which would indicate it contains alignment data. If such a line 
/// is found, the function will return `Ok(())`, signaling the file does contain alignments.
///
/// If the end of the file is reached without encountering any alignment data (i.e., all 
/// lines start with &#39;@&#39;), an error is returned with the message &quot;SAM file is empty&quot;.
///
/// # Arguments
/// * `sam_file` - A string slice that holds the path to the SAM file.
///
/// # Returns
/// * `Ok(())` if the SAM file contains alignment data.
/// * `Err(io::Error)` if the SAM file does not contain alignment data or if there was an 
/// error reading the file.
///
/// # Errors
/// This function will return an error if there was an issue reading the file or if the 
/// file does not contain alignment data.
///
/// # Example
/// ```no_run
/// let result = is_sam_file_empty(&quot;/path/to/your/file.sam&quot;);
/// match result {
///     Ok(_) =&gt; println!(&quot;The SAM file contains alignments.&quot;),
///     Err(e) =&gt; eprintln!(&quot;An error occurred: {}&quot;, e),
/// }
/// ```
</span><span class="kw">pub fn </span>is_sam_file_empty(sam_file: <span class="kw-2">&amp;</span>str) -&gt; <span class="prelude-ty">Result</span>&lt;(), io::Error&gt; {
    <span class="kw">let </span>file = File::open(sam_file)<span class="question-mark">?</span>;
    <span class="kw">let </span>reader = BufReader::new(file);

    <span class="kw">for </span>line <span class="kw">in </span>reader.lines() {
        <span class="kw">let </span>line = line<span class="question-mark">?</span>;
        <span class="kw">if </span>!line.starts_with(<span class="string">&#39;@&#39;</span>) {
            <span class="kw">return </span><span class="prelude-val">Ok</span>(());
        }
    }

    <span class="prelude-val">Err</span>(io::Error::new(io::ErrorKind::InvalidData, <span class="string">&quot;SAM file is empty&quot;</span>))
}

<span class="doccomment">/// Perform alignment of trimmed fastq files to the target RNA bowtie2 reference to calculate
/// the number of unique RNA present in each sample. UMI error correction and deduplication
/// is also done. Only aligned reads are output to SAM file in order to speed up UMI deduplication.
///
/// Returns a Vector of paths to SAM files containing RNA counts for each sample.
///
/// # Arguments
/// * `trimmed_fastqs` - Vector of paths to trimmed fastq files.
/// * `library_type` - HashMap mapping sample names to their library types.
/// * `sample_names` - Vector of sample names.
/// * `config` - Struct containing user-specified analysis configurations, see below.
///     * `reference` - Path to the bowtie2 reference.
///     * `num_threads` - Number of threads to use for alignment and conversion processes.
///     * `include_unaligned_reads` - Boolean flag for whether to include unaligned reads or not.
///     * `write_metrics` - Boolean flag for whether to write sample metrics to an output file or not.
///
/// # Example
/// ```
/// let trimmed_fastqs = vec![&quot;sample1.cut.fastq&quot;, &quot;sample2.cut.fastq&quot;, &quot;sample3.cut.fastq&quot;];
/// let sample_names = vec![&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;];
/// let config = Config::new(matches);
///
/// rna_discovery_calculation(trimmed_fastqs, sample_names, config);
/// // This will generate &quot;sample1.miRNA.sam&quot;, &quot;sample2.miRNA.sam&quot;, &quot;sample3.miRNA.sam&quot; files
/// // and will return a vector of SAM file paths.
/// ```
</span><span class="kw">pub fn </span>rna_discovery_calculation(
                                trimmed_fastqs: Vec&lt;String&gt;, 
                                sample_names: Vec&lt;String&gt;,
                                config: <span class="kw-2">&amp;</span>Config
                            ) -&gt; <span class="prelude-ty">Result</span>&lt;Vec&lt;String&gt;, io::Error&gt; {
    <span class="comment">// Calculate the number of reads in each file if metrics are enabled
    </span><span class="kw">let </span>read_counts = 
    <span class="kw">if </span>config.write_metrics {
        <span class="kw">let </span>fastq_files: Vec&lt;String&gt; = capture_target_files(<span class="string">&quot;_R1_001.fastq.gz&quot;</span>);
        <span class="kw">match </span>get_read_counts(fastq_files, <span class="kw-2">&amp;</span>sample_names) {
            <span class="prelude-val">Ok</span>(read_counts) =&gt; read_counts,
            <span class="prelude-val">Err</span>(err) =&gt; <span class="macro">panic!</span>(<span class="string">&quot;An error occurred while getting read counts: {}&quot;</span>, err),
        }
    } <span class="kw">else </span>{
        HashMap::new() <span class="comment">// Return an empty HashMap when write_metrics is not true
    </span>};
    
    <span class="kw">let </span>num_of_fastqs = trimmed_fastqs.len() <span class="kw">as </span>u64;
    <span class="kw">let </span>progress_br = ProgressBar::new(num_of_fastqs);
    
    progress_br.set_style(
        ProgressStyle::default_bar()
            .template(<span class="string">&quot;{spinner:.green} [{elapsed_precise}] [{bar:50.cyan/blue}] {pos}/{len} {msg} ({percent}%)&quot;</span>)
                .expect(<span class="string">&quot;Progress bar error&quot;</span>)
            .progress_chars(<span class="string">&quot;#&gt;-&quot;</span>),
    );

    <span class="comment">// Set analysis reference, i.e., RNA type name, based on the user provided reference name
    </span><span class="kw">let </span>reference_name = Path::new(<span class="kw-2">&amp;</span>config.alignment_reference).file_name().unwrap().to_str().unwrap();
    progress_br.set_message(<span class="macro">format!</span>(<span class="string">&quot;Calculating the number of {} present...&quot;</span>, reference_name));

    <span class="comment">// Set bowtie2 flag to keep unaligned reads
    </span><span class="kw">let </span>unaligned = 
    <span class="kw">if </span>config.include_unaligned_reads {
        <span class="string">&quot;&quot;
    </span>} <span class="kw">else </span>{
        <span class="string">&quot;--no-unal&quot;
    </span>};

    <span class="comment">// Create file for writing metrics if necessary
    </span><span class="kw">let </span><span class="kw-2">mut </span>writer: <span class="prelude-ty">Option</span>&lt;csv_writer&lt;File&gt;&gt; = <span class="prelude-val">None</span>;
    <span class="kw">if </span>config.write_metrics {
        writer = <span class="prelude-val">Some</span>(csv_writer::from_path(<span class="string">&quot;metrics.csv&quot;</span>)<span class="question-mark">?</span>);
        writer.as_mut().unwrap().write_record(<span class="kw-2">&amp;</span>[<span class="string">&quot;sample&quot;</span>, <span class="string">&quot;read count&quot;</span>, <span class="string">&quot;quality score&quot;</span>, reference_name])<span class="question-mark">?</span>;
    }

    <span class="kw">for </span>(fastq_file, sample_name) <span class="kw">in </span>trimmed_fastqs.iter().zip(sample_names.iter()) {
        <span class="comment">// Align to bowtie2 reference
        // Make mapped SAM file with header for faster analysis
        </span><span class="kw">let _ </span>= Command::new(<span class="string">&quot;bowtie2&quot;</span>)
            .args(<span class="kw-2">&amp;</span>[<span class="string">&quot;--norc&quot;
                    </span>, <span class="string">&quot;--threads&quot;
                    </span>, <span class="kw-2">&amp;</span>config.num_threads.to_string()
                    , unaligned
                    , <span class="string">&quot;-x&quot;
                    </span>, <span class="kw-2">&amp;</span>config.alignment_reference
                    , <span class="string">&quot;-U&quot;
                    </span>, fastq_file
                    , <span class="string">&quot;-S&quot;
                    </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.sam&quot;</span>, sample_name)]
                )
            .output()
            .expect(<span class="string">&quot;Failed to run bowtie2&quot;</span>);

        <span class="kw">let </span>check_result = is_sam_file_empty(<span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.sam&quot;</span>, sample_name));

        <span class="comment">// Skip analysis of sample if no alignments are found
        </span><span class="kw">match </span>check_result {
            <span class="prelude-val">Ok</span>(<span class="kw">_</span>) =&gt; {
                <span class="comment">// Continue processing this sample
            </span>}
            <span class="prelude-val">Err</span>(e) =&gt; {
                <span class="macro">println!</span>(<span class="string">&quot;Error while checking {}: {}. Please check input samples and parameters.&quot;
                        </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.sam&quot;</span>, sample_name)
                        , e);
                <span class="comment">// Skip this sample
                </span><span class="kw">continue</span>;
            }
        }

        <span class="comment">// Can use SAM file for deduplication, but a BAM file uses less memory
        </span><span class="kw">let _ </span>= Command::new(<span class="string">&quot;samtools&quot;</span>)
            .args(<span class="kw-2">&amp;</span>[<span class="string">&quot;view&quot;
                    </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;-@ {}&quot;</span>, <span class="kw-2">&amp;</span>config.num_threads).to_string()
                    , <span class="string">&quot;--with-header&quot;
                    </span>, <span class="string">&quot;-o&quot;
                    </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.bam&quot;</span>, sample_name)
                    , <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.sam&quot;</span>, sample_name)]
                )
            .output()
            .expect(<span class="string">&quot;Failed to run samtools to convert SAM to BAM&quot;</span>);

        <span class="comment">// Error correct UMIs in sample bam file
        </span><span class="kw">let </span>repr_umis: HashSet&lt;Vec&lt;u8&gt;&gt; = find_true_umis(<span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.bam&quot;</span>, sample_name), config.levenshtein_distance)<span class="question-mark">?</span>;

        <span class="comment">// Deduplicate mapped bam file
        </span><span class="kw">let </span>deduplication_result = deduplicate_bam(<span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.UMI.bam&quot;</span>, sample_name)
                                                                                , sample_name
                                                                                , repr_umis);

        <span class="comment">// Check the result
        </span><span class="kw">match </span>deduplication_result {
            <span class="prelude-val">Ok</span>(<span class="kw">_</span>) =&gt; (),
            <span class="prelude-val">Err</span>(e) =&gt; <span class="macro">eprintln!</span>(<span class="string">&quot;An error occurred while deduplicating the sample {}: {}&quot;</span>, sample_name, e),
        }

        <span class="comment">// Convert deduplicated bam file to a sam file
        </span><span class="kw">let _ </span>= Command::new(<span class="string">&quot;samtools&quot;</span>)
            .args(<span class="kw-2">&amp;</span>[<span class="string">&quot;view&quot;
                    </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;-@ {}&quot;</span>, <span class="kw-2">&amp;</span>config.num_threads).to_string()
                    , <span class="string">&quot;--with-header&quot;
                    </span>, <span class="string">&quot;-o&quot;
                    </span>, <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.{}.sam&quot;</span>, sample_name, reference_name)
                    , <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.dedup.bam&quot;</span>, sample_name)]
                )
            .output()
            .expect(<span class="string">&quot;Failed to run samtools to convert BAM to SAM&quot;</span>);

        <span class="comment">// Generate data for the specific RNAs captured
        </span><span class="kw">let _  </span>= generate_rna_counts(<span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.{}.sam&quot;</span>, sample_name, reference_name), sample_name, reference_name);

        <span class="comment">// Set correct read count depending on if subsampling was performed
        </span><span class="kw">if </span>config.write_metrics {
            <span class="kw">let </span>count: u64;
            <span class="kw">if let </span><span class="prelude-val">Some</span>(n_reads) = config.subsample_fastqs {
                count = n_reads <span class="kw">as </span>u64 <span class="comment">// User chose the number of reads to subsample to
            </span>} <span class="kw">else if </span>config.subsample_to_lowest {
                count = <span class="kw-2">*</span>read_counts.values().min().unwrap() <span class="comment">// Count equals fastq with lowest read count
            </span>} <span class="kw">else </span>{
                count = read_counts[sample_name] <span class="comment">// Total read count
            </span>};

            <span class="comment">// Calculate the percentage of aligned reads
            </span><span class="kw">if let </span><span class="prelude-val">Ok</span>(percent_alignment) = get_percent_alignment(sample_name, config.num_threads, reference_name, count) {
                <span class="kw">let </span>avg_quality = average_read_quality(fastq_file).expect(<span class="string">&quot;Failed to calculate read quality&quot;</span>);

                writer
                    .as_mut()
                    .expect(<span class="string">&quot;Failed to write record&quot;</span>)
                    .write_record(<span class="kw-2">&amp;</span>[sample_name, <span class="kw-2">&amp;</span>count.to_string(), <span class="kw-2">&amp;</span>avg_quality.to_string(), <span class="kw-2">&amp;</span>percent_alignment.to_string()])<span class="question-mark">?</span>;
            };
        }

        progress_br.inc(<span class="number">1</span>);
    }

    <span class="kw">if </span>config.write_metrics {
        writer.expect(<span class="string">&quot;Failed to flush writer&quot;</span>).flush()<span class="question-mark">?</span>;
    }

    <span class="kw">let </span>sam_files: Vec&lt;String&gt; = capture_target_files(<span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;.{}.sam&quot;</span>, reference_name));

    <span class="comment">// Stop analysis if all samples lack alignment data 
    </span><span class="kw">if </span>sam_files.len() == <span class="number">0 </span>{
        <span class="macro">panic!</span>(<span class="string">&quot;No deduplicated sam files found. Analysis aborted.&quot;</span>);
    }

    progress_br.finish_with_message(<span class="macro">format!</span>(<span class="string">&quot;Finished counting {}&quot;</span>, reference_name));

    <span class="prelude-val">Ok</span>(sam_files)
}

<span class="doccomment">/// This function counts the occurrence of each RNA in an input SAM file, computes their 
/// reads per million (RPM),and writes this information to an output CSV file. The CSV 
/// file will contain the name of each unique RNA, its count,and RPM for the given sample.
///
/// The output CSV file will have the format `&quot;{sample_name}_{reference_name}_counts.csv&quot;` 
/// and will be located in the directory where the program is run.
///
/// # Returns
///
/// * A `Result` which is `Ok` if the function executed successfully. It is `Err` if an error occurred, 
///   with a description of the error.
///
/// # Arguments
///
/// * `input_sam_file`: The path to the input SAM file.
/// * `sample_name`: The name of the sample.
/// * `reference_name`: The name of the bowtie2 reference.
/// # Example
/// ```
/// let sam_file = &quot;sample1.sam&quot;;
/// let reference_name = &quot;miRNA&quot;
/// 
/// let _ = generate_rna_counts(sam_file, &quot;sample1&quot;, reference_name);
/// // This will generate a file &quot;sample1_miRNA_counts.csv&quot; with each miRNA found and
/// // their counts and CPM.
/// ```
</span><span class="kw">pub fn </span>generate_rna_counts(input_sam_file: <span class="kw-2">&amp;</span>str, sample_name: <span class="kw-2">&amp;</span>String, reference_name: <span class="kw-2">&amp;</span>str) 
    -&gt; <span class="prelude-ty">Result</span>&lt;(), Box&lt;<span class="kw">dyn </span>std::error::Error&gt;&gt; {
    <span class="kw">let </span>file = File::open(input_sam_file)<span class="question-mark">?</span>;
    <span class="kw">let </span>reader = BufReader::new(file);

    <span class="comment">// Initialize a HashMap to count the occurrence of each RNA
    </span><span class="kw">let </span><span class="kw-2">mut </span>rna_counts: HashMap&lt;String, i32&gt; = HashMap::new();

    <span class="comment">// Read the SAM file line by line
    </span><span class="kw">for </span>line <span class="kw">in </span>reader.lines() {
        <span class="kw">let </span>line = line<span class="question-mark">?</span>;

        <span class="comment">// Skip the header lines
        </span><span class="kw">if </span>line.starts_with(<span class="string">&#39;@&#39;</span>) {
            <span class="kw">continue</span>;
        }

        <span class="comment">// Split each line into fields and extract the sequence name
        </span><span class="kw">let </span>fields: Vec&lt;<span class="kw">_</span>&gt; = line.split(<span class="string">&#39;\t&#39;</span>).collect();
        <span class="kw">if </span>fields.len() &lt; <span class="number">3 </span>{
            <span class="kw">continue</span>;
        }
        <span class="kw">let </span>ref_seq_name = fields[<span class="number">2</span>];

        <span class="comment">// Increment the count for this RNA in the HashMap
        </span><span class="kw-2">*</span>rna_counts.entry(ref_seq_name.to_string()).or_insert(<span class="number">0</span>) += <span class="number">1</span>;
    }

    <span class="comment">// Calculate the total number of mapped reads
    </span><span class="kw">let </span>total_mapped_reads: i32 = rna_counts.values().sum();

    <span class="comment">// Prepare vectors for DataFrame creation
    </span><span class="kw">let </span><span class="kw-2">mut </span>names: Vec&lt;String&gt; = Vec::new();
    <span class="kw">let </span><span class="kw-2">mut </span>counts: Vec&lt;i32&gt; = Vec::new();
    <span class="kw">let </span><span class="kw-2">mut </span>rpms: Vec&lt;f64&gt; = Vec::new();

    <span class="comment">// For each RNA, calculate its RPM and store its name, count, and RPM
    </span><span class="kw">for </span>(name, count) <span class="kw">in </span>rna_counts {
        names.push(name);
        counts.push(count);
        rpms.push((count <span class="kw">as </span>f64 / total_mapped_reads <span class="kw">as </span>f64) * <span class="number">1_000_000.0</span>); <span class="comment">// Calculate RPM
    </span>}

    <span class="comment">// Create a DataFrame with &#39;name&#39;, &#39;count&#39;, and &#39;RPM&#39; as columns
    </span><span class="kw">let </span><span class="kw-2">mut </span>counts_df = DataFrame::new(<span class="macro">vec!</span>[
        Series::new(<span class="string">&quot;name&quot;</span>, names),
        Series::new(<span class="string">&quot;count&quot;</span>, counts),
        Series::new(<span class="string">&quot;RPM&quot;</span>, rpms),
    ])<span class="question-mark">?</span>;

    <span class="comment">// Sort the DataFrame in-place by the &#39;count&#39; column
    </span><span class="kw">let _ </span>= counts_df.sort_in_place(<span class="kw-2">&amp;</span>[<span class="string">&quot;count&quot;</span>], <span class="bool-val">true</span>);

    <span class="comment">// Create a new CSV file to store the DataFrame
    </span><span class="kw">let </span>file = File::create(<span class="macro">format!</span>(<span class="string">&quot;{}_{}_counts.csv&quot;</span>, sample_name, reference_name)).expect(<span class="string">&quot;could not create file&quot;</span>);

    <span class="comment">// Write the DataFrame to the CSV file with headers
    </span>CsvWriter::new(file)
        .has_header(<span class="bool-val">true</span>)
        .finish(<span class="kw-2">&amp;mut </span>counts_df)
        .unwrap();

    <span class="prelude-val">Ok</span>(())
}

<span class="doccomment">/// Get the percentage of reads that aligned to the specified reference RNA type.
///
/// Divide reads aligned with total reads to calculate percent alignment.  
///
/// /// # Arguments
///
/// * `count` - The total number of reads analyzed. 
/// * `sample_name` - The name of the sample.
/// * `reference_name` - Path to the bowtie2 reference.
/// * `num_threads` - Number of threads to use for alignment and conversion processes.
/// 
/// # Returns
///
/// The percent each fastq aligned to the specified RNA type.
/// 
/// # Example
/// ```
/// let reference_name = &quot;miRNA&quot;
/// let count = 12345678
/// 
/// let percent_alignment = get_percent_alignment(&quot;sample1&quot;, 4, reference_name, count);
/// println!(&quot;The fastq contains {} percent {}.&quot;, percent_alignment, reference_name);
/// ```
</span><span class="kw">pub fn </span>get_percent_alignment(
                    sample_name: <span class="kw-2">&amp;</span>String,
                    num_threads: u8,
                    reference_name: <span class="kw-2">&amp;</span>str,
                    count: u64
                ) -&gt; <span class="prelude-ty">Result</span>&lt;f64, Box&lt;<span class="kw">dyn </span>std::error::Error&gt;&gt; {
    <span class="comment">// Get the number of RNA alignments in the input fastq
    </span><span class="kw">let </span>output_result = Command::new(<span class="string">&quot;samtools&quot;</span>)
        .args(<span class="kw-2">&amp;</span>[
            <span class="string">&quot;view&quot;</span>,
            <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;-@ {}&quot;</span>, <span class="kw-2">&amp;</span>num_threads),
            <span class="string">&quot;-c&quot;</span>,
            <span class="string">&quot;-F 4&quot;</span>,
            <span class="kw-2">&amp;</span><span class="macro">format!</span>(<span class="string">&quot;{}.{}.sam&quot;</span>, sample_name, reference_name),
        ])
        .output();

    <span class="kw">let </span>output = output_result.expect(<span class="string">&quot;Failed to execute command&quot;</span>);
    <span class="kw">let </span>stdout = <span class="kw">match </span>String::from_utf8(output.stdout) {
        <span class="prelude-val">Ok</span>(v) =&gt; v,
        <span class="prelude-val">Err</span>(e) =&gt; <span class="macro">panic!</span>(<span class="string">&quot;Invalid UTF-8 sequence: {}&quot;</span>, e),
    };

    <span class="kw">let </span>parsed = <span class="kw">match </span>stdout.trim().parse::&lt;f64&gt;() {
        <span class="prelude-val">Ok</span>(v) =&gt; v,
        <span class="prelude-val">Err</span>(e) =&gt; { 
            <span class="macro">eprintln!</span>(<span class="string">&quot;Failed to parse output to f64: {:?}&quot;</span>, e);
            <span class="number">0.0 </span><span class="comment">// Set read count to zero if there&#39;s no output
        </span>},
    };
    <span class="kw">let </span>percent_alignment = <span class="number">100.0 </span>* parsed / count <span class="kw">as </span>f64;

    <span class="prelude-val">Ok</span>(percent_alignment)
}
</code></pre></div></section></main></body></html>